src/group_theory/solvable.lean

```diff
@@ -0,0 +1,129 @@
+/-
+Copyright (c) 2020 Jordan Brown, Thomas Browning and Patrick Lutz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jordan Brown, Thomas Browning and Patrick Lutz
+-/
+
+import group_theory.abelianization
+
+/-!
+# Solvable Groups
+
+In this file we introduce the notion of a solvable group. We define a solvable group as one whose
+nth commutator is `1` for some `n`. This requires defining the commutator of two subgroups and
+the derived series of a group.
+
+## Main definitions
+
+* `general_commutator H₁ H₂` : the commutator of the subgroups `H₁` and `H₂`
+* `derived_series G n` : the `n`th term in the derived series of `G`, defined by iterating
+  `general_commutator` starting with the top subgroup
+* `is_solvable G` : the group `G` is solvable
+-/
+
+open subgroup
+
+variables {G : Type*} [group G]
+
+section general_commutator
+
+/-- The commutator of two subgroups `H₁` and `H₂`. -/
+def general_commutator (H₁ : subgroup G) (H₂ : subgroup G) : subgroup G :=
+subgroup.closure {x | ∃ (p ∈ H₁) (q ∈ H₂), p * q * p⁻¹ * q⁻¹ = x}
+
+instance general_commutator_normal (H₁ : subgroup G) (H₂ : subgroup G) [h₁ : H₁.normal]
+  [h₂ : H₂.normal] : normal (general_commutator H₁ H₂) :=
+begin
+  let base : set G := {x | ∃ (p ∈ H₁) (q ∈ H₂), p * q * p⁻¹ * q⁻¹ = x},
+  suffices h_base : base = group.conjugates_of_set base,
+  { dsimp only [general_commutator, ←base],
+    rw h_base,
+    exact subgroup.normal_closure_normal },
+  apply set.subset.antisymm group.subset_conjugates_of_set,
+  intros a h,
+  rw group.mem_conjugates_of_set_iff at h,
+  rcases h with ⟨b, ⟨c, hc, e, he, rfl⟩, d, rfl⟩,
+  exact ⟨d * c * d⁻¹, h₁.conj_mem c hc d, d * e * d⁻¹, h₂.conj_mem e he d, by group⟩,
+end
+
+lemma general_commutator_mono {H₁ H₂ K₁ K₂ : subgroup G} (h₁ : H₁ ≤ K₁) (h₂ : H₂ ≤ K₂) :
+  general_commutator H₁ H₂ ≤ general_commutator K₁ K₂ :=
+begin
+  apply closure_mono,
+  rintros x ⟨p, hp, q, hq, rfl⟩,
+  exact ⟨p, h₁ hp, q, h₂ hq, rfl⟩,
+end
+
+lemma general_commutator_eq_normal_closure_self (H₁ : subgroup G) (H₂ : subgroup G) [H₁.normal]
+  [H₂.normal] : general_commutator H₁ H₂ = normal_closure (general_commutator H₁ H₂) :=
+eq.symm normal_closure_eq_self_of_normal
+
+lemma general_commutator_def' (H₁ H₂ : subgroup G) [H₁.normal] [H₂.normal] :
+  general_commutator H₁ H₂ = normal_closure {x | ∃ (p ∈ H₁) (q ∈ H₂), p * q * p⁻¹ * q⁻¹ = x} :=
+by rw [general_commutator_eq_normal_closure_self, general_commutator,
+  normal_closure_closure_eq_normal_closure]
+
+end general_commutator
+
+section derived_series
+
+variables (G)
+
+/-- The derived series of the group `G`, obtained by starting from the subgroup `⊤` and repeatedly
+  taking the commutator of the previous subgroup with itself for `n` times. -/
+def derived_series (n : ℕ) : subgroup G :=
+nat.rec_on n (⊤ : subgroup G) (λ _ H, general_commutator H H)
```

Comment:
```suggestion
| 0       := (⊤ : subgroup G)
| (n + 1) := general_commutator (derived_series n) (derived_series n)
```
I prefer this, because I find it more readable.

Reply:
You should even be able to get away without the type ascription for `⊤`, i.e.,
```lean
| 0       := ⊤
| (n + 1) := general_commutator (derived_series n) (derived_series n)
```

Reply:
I had to change the statement of the definition to `def derived_series : Π n : ℕ, subgroup G` to get it to compile. I hope that's okay.

Reply:
Okay, this is resolved now (see comments below).